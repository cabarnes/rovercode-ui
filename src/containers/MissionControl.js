import React from 'react';
import {
  Box,
  Button,
  Divider,
  Drawer,
  ExpansionPanel,
  ExpansionPanelSummary,
  ExpansionPanelDetails,
  Grid,
  IconButton,
  List,
  ListItem,
  ListSubheader,
  Typography,
} from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import { ExpandMore, Close } from '@material-ui/icons';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import PropTypes from 'prop-types';

import CodeViewer from '@/components/CodeViewer';
import Console from '@/components/Console';
import Control from '@/components/Control';
import Indicator from '@/components/Indicator';
import ProgramName from '@/components/ProgramName';
import ProgramTags from '@/components/ProgramTags';
import Workspace from '@/components/Workspace';

const mapStateToProps = ({ code }) => ({ code });

const MissionControl = ({ location, code }) => {
  const [state, setState] = React.useState({
    open: false,
  });

  const handleOnClose = () => setState({ ...state, open: false });

  const WideBox = withStyles(() => ({
    root: {
      width: '100%',
    },
  }))(Box);

  return (
    <>
      <Drawer anchor="left" open={state.open} onClose={handleOnClose}>
        <List
          subheader={(
            <ListSubheader component="div" id="nested-list-subheader">
              <Grid container direction="row" justify="space-between" alignItems="center">
                <Grid item>
                  <Typography variant="h6">
                    <FormattedMessage
                      id="app.mission_control.details"
                      description="Header for project details"
                      defaultMessage="Project Details"
                    />
                  </Typography>
                </Grid>
                <Grid item>
                  <IconButton onClick={handleOnClose}>
                    <Close />
                  </IconButton>
                </Grid>
              </Grid>
            </ListSubheader>
          )}
        >
          <Divider />
          <ListItem>
            <ProgramName location={location} />
          </ListItem>
          <ListItem>
            <ProgramTags />
          </ListItem>
        </List>
      </Drawer>

      <Grid container direction="row" justify="space-evenly" alignItems="flex-start" spacing={0}>
        <Grid item container xs={10} direction="column" justify="center" alignItems="stretch" spacing={2}>
          <Grid item container direction="row" justify="space-between">
            <Grid item>
              <Box m={1}>
                <Typography variant="h6">
                  {code.name}
                </Typography>
              </Box>
            </Grid>
            <Grid item>
              <Button color="default" onClick={() => setState({ ...state, open: true })}>
                <FormattedMessage
                  id="app.mission_control.edit"
                  description="Button to edit the project"
                  defaultMessage="Edit Project"
                />
              </Button>
            </Grid>
          </Grid>
          <Grid item>
            <Workspace location={location}>
              <Control />
            </Workspace>
          </Grid>
        </Grid>
        <Grid item container xs={2} direction="column" alignItems="stretch" spacing={2}>
          <Grid item>
            <ExpansionPanel>
              <ExpansionPanelSummary expandIcon={<ExpandMore />}>
                <Typography>
                  <FormattedMessage
                    id="app.mission_control.sensors"
                    description="Header for list of the user's sensors"
                    defaultMessage="Sensors"
                  />
                </Typography>
              </ExpansionPanelSummary>
              <ExpansionPanelDetails>
                <WideBox p={2} border={1} borderRadius="borderRadius" borderColor="grey.500">
                  <Indicator />
                </WideBox>
              </ExpansionPanelDetails>
            </ExpansionPanel>
          </Grid>
          <Grid item>
            <ExpansionPanel>
              <ExpansionPanelSummary expandIcon={<ExpandMore />}>
                <Typography>
                  <FormattedMessage
                    id="app.mission_control.console"
                    description="Header for debug console"
                    defaultMessage="Debug Console"
                  />
                </Typography>
              </ExpansionPanelSummary>
              <ExpansionPanelDetails>
                <Console />
              </ExpansionPanelDetails>
            </ExpansionPanel>
          </Grid>
          <Grid item>
            <CodeViewer>
              <FormattedMessage
                id="app.mission_control.show_code"
                description="Button label for displaying user's code"
                defaultMessage="View JavaScript"
              />
            </CodeViewer>
          </Grid>
        </Grid>
      </Grid>
    </>
  );
};

MissionControl.defaultProps = {
  location: {
    state: {
      readOnly: false,
    },
  },
};

MissionControl.propTypes = {
  location: PropTypes.shape({
    state: PropTypes.shape({
      readOnly: PropTypes.bool,
    }),
  }),
  code: PropTypes.shape({
    name: PropTypes.string,
  }).isRequired,
};

export default hot(module)(connect(mapStateToProps)(MissionControl));
