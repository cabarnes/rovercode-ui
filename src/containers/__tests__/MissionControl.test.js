import React from 'react';
import { MemoryRouter } from 'react-router';
import { Provider as ReduxProvider } from 'react-redux';
import { createStore } from 'redux';
import { Button, Drawer } from '@material-ui/core';
import { Cookies } from 'react-cookie';
import PropTypes from 'prop-types';
import rootReducer from '@/reducers/index';
import MissionControl from '../MissionControl';

jest.mock('@/components/CodeViewer', () => () => <div />);
jest.mock('@/components/Console', () => () => <div />);
jest.mock('@/components/Control', () => () => <div />);
jest.mock('@/components/Indicator', () => () => <div />);
jest.mock('@/components/ProgramName', () => () => <div />);
jest.mock('@/components/Workspace', () => () => <div />);

const cookiesValues = { auth_jwt: '1234' };
const cookies = new Cookies(cookiesValues);

describe('The MissionControl container', () => {
  let store;
  let wrapper;
  beforeEach(() => {
    store = createStore(rootReducer, {
      code: {
        jsCode: 'testcode',
        tags: [],
      },
      rover: {
        isFetching: false,
        rovers: {
          next: null,
          previous: null,
          results: [],
        },
        commands: [],
      },
    });
    store.dispatch = jest.fn(() => Promise.resolve({}));
  });

  test('renders on the page with no errors', () => {
    const context = { cookies };
    wrapper = mountWithIntl(
      <ReduxProvider store={store}>
        <MemoryRouter>
          <MissionControl store={store} />
        </MemoryRouter>
      </ReduxProvider>, {
        context,
        childContextTypes: { cookies: PropTypes.instanceOf(Cookies) },
      },
    );
    const mcWrapper = wrapper.find(MissionControl);

    expect(mcWrapper).toMatchSnapshot();
  });

  test('shows and dismisses drawer on edit', () => {
    const context = { cookies };
    wrapper = shallowWithIntl(
      <ReduxProvider store={store}>
        <MemoryRouter>
          <MissionControl store={store} />
        </MemoryRouter>
      </ReduxProvider>, {
        context,
        childContextTypes: { cookies: PropTypes.instanceOf(Cookies) },
      },
    ).dive().dive().dive()
      .dive()
      .dive()
      .dive()
      .dive();

    expect(wrapper.find(Drawer).prop('open')).toBe(false);

    wrapper.find(Button).simulate('click');
    wrapper.update();

    expect(wrapper.find(Drawer).prop('open')).toBe(true);

    wrapper.find(Drawer).simulate('close');
    wrapper.update();

    expect(wrapper.find(Drawer).prop('open')).toBe(false);
  });
});
