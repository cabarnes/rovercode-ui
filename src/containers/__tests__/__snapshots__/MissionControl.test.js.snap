// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The MissionControl container renders on the page with no errors 1`] = `
<MissionControl
  location={
    Object {
      "state": Object {
        "readOnly": false,
      },
    }
  }
  store={
    Object {
      "dispatch": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "payload": Promise {},
              "type": "FETCH_TAGS",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Grid>
    <div
      className="ui grid"
    >
      <GridRow>
        <div
          className="row"
        >
          <GridColumn
            width={3}
          >
            <div
              className="three wide column"
            >
              <GridRow>
                <div
                  className="row"
                >
                  <Segment
                    basic={true}
                    compact={true}
                  >
                    <div
                      className="ui basic compact segment"
                    >
                      <Component
                        location={
                          Object {
                            "state": Object {
                              "readOnly": false,
                            },
                          }
                        }
                      >
                        <div />
                      </Component>
                    </div>
                  </Segment>
                </div>
              </GridRow>
              <Divider>
                <div
                  className="ui divider"
                />
              </Divider>
              <GridRow>
                <div
                  className="row"
                >
                  <Header
                    as="h2"
                    textAlign="center"
                  >
                    <h2
                      className="ui center aligned header"
                    >
                      <FormattedMessage
                        defaultMessage="Tags"
                        id="app.mission_control.tags"
                        values={Object {}}
                      >
                        <span>
                          Tags
                        </span>
                      </FormattedMessage>
                    </h2>
                  </Header>
                  <withCookies(Component)>
                    <InjectIntl(Connect(ProgramTags))
                      allCookies={Object {}}
                      cookies={
                        Cookies {
                          "HAS_DOCUMENT_COOKIE": true,
                          "changeListeners": Array [
                            [Function],
                          ],
                          "cookies": Object {},
                          "hooks": undefined,
                        }
                      }
                    >
                      <Connect(ProgramTags)
                        allCookies={Object {}}
                        cookies={
                          Cookies {
                            "HAS_DOCUMENT_COOKIE": true,
                            "changeListeners": Array [
                              [Function],
                            ],
                            "cookies": Object {},
                            "hooks": undefined,
                          }
                        }
                        intl={
                          Object {
                            "defaultFormats": Object {},
                            "defaultLocale": "en",
                            "formatDate": [Function],
                            "formatHTMLMessage": [Function],
                            "formatMessage": [Function],
                            "formatNumber": [Function],
                            "formatPlural": [Function],
                            "formatRelative": [Function],
                            "formatTime": [Function],
                            "formats": Object {},
                            "formatters": Object {
                              "getDateTimeFormat": [Function],
                              "getMessageFormat": [Function],
                              "getNumberFormat": [Function],
                              "getPluralFormat": [Function],
                              "getRelativeFormat": [Function],
                            },
                            "locale": "en",
                            "messages": Object {},
                            "now": [Function],
                            "onError": [Function],
                            "textComponent": "span",
                            "timeZone": null,
                          }
                        }
                      >
                        <ProgramTags
                          allCookies={Object {}}
                          changeProgramTags={[Function]}
                          code={
                            Object {
                              "jsCode": "testcode",
                              "tags": Array [],
                            }
                          }
                          cookies={
                            Cookies {
                              "HAS_DOCUMENT_COOKIE": true,
                              "changeListeners": Array [
                                [Function],
                              ],
                              "cookies": Object {},
                              "hooks": undefined,
                            }
                          }
                          fetchTags={[Function]}
                          intl={
                            Object {
                              "defaultFormats": Object {},
                              "defaultLocale": "en",
                              "formatDate": [Function],
                              "formatHTMLMessage": [Function],
                              "formatMessage": [Function],
                              "formatNumber": [Function],
                              "formatPlural": [Function],
                              "formatRelative": [Function],
                              "formatTime": [Function],
                              "formats": Object {},
                              "formatters": Object {
                                "getDateTimeFormat": [Function],
                                "getMessageFormat": [Function],
                                "getNumberFormat": [Function],
                                "getPluralFormat": [Function],
                                "getRelativeFormat": [Function],
                              },
                              "locale": "en",
                              "messages": Object {},
                              "now": [Function],
                              "onError": [Function],
                              "textComponent": "span",
                              "timeZone": null,
                            }
                          }
                          tag={
                            Object {
                              "error": null,
                              "isFetching": false,
                              "tags": Array [],
                            }
                          }
                        >
                          <Grid
                            centered={true}
                            container={true}
                          >
                            <div
                              className="ui centered container grid"
                            >
                              <GridRow>
                                <div
                                  className="row"
                                >
                                  <Dropdown
                                    additionLabel="Add "
                                    additionPosition="top"
                                    allowAdditions={true}
                                    closeOnBlur={true}
                                    closeOnChange={true}
                                    closeOnEscape={true}
                                    deburr={false}
                                    fluid={true}
                                    icon="dropdown"
                                    minCharacters={1}
                                    multiple={true}
                                    noResultsMessage="No results found."
                                    onAddItem={[Function]}
                                    onChange={[Function]}
                                    openOnFocus={true}
                                    options={Array []}
                                    placeholder="Add tags..."
                                    renderLabel={[Function]}
                                    search={true}
                                    searchInput="text"
                                    selectOnBlur={true}
                                    selectOnNavigation={true}
                                    selection={true}
                                    value={Array []}
                                    wrapSelection={true}
                                  >
                                    <Ref
                                      innerRef={
                                        Object {
                                          "current": <div
                                            aria-expanded="false"
                                            class="ui fluid multiple search selection dropdown"
                                            role="combobox"
                                          >
                                            <input
                                              aria-autocomplete="list"
                                              autocomplete="off"
                                              class="search"
                                              tabindex="0"
                                              type="text"
                                              value=""
                                            />
                                            <span
                                              class="sizer"
                                            />
                                            <div
                                              aria-atomic="true"
                                              aria-live="polite"
                                              class="default text"
                                              role="alert"
                                            >
                                              Add tags...
                                            </div>
                                            <i
                                              aria-hidden="true"
                                              class="dropdown icon"
                                            />
                                            <div
                                              aria-multiselectable="true"
                                              class="menu transition"
                                              role="listbox"
                                            >
                                              <div
                                                class="message"
                                              >
                                                No results found.
                                              </div>
                                            </div>
                                          </div>,
                                        }
                                      }
                                    >
                                      <RefFindNode
                                        innerRef={
                                          Object {
                                            "current": <div
                                              aria-expanded="false"
                                              class="ui fluid multiple search selection dropdown"
                                              role="combobox"
                                            >
                                              <input
                                                aria-autocomplete="list"
                                                autocomplete="off"
                                                class="search"
                                                tabindex="0"
                                                type="text"
                                                value=""
                                              />
                                              <span
                                                class="sizer"
                                              />
                                              <div
                                                aria-atomic="true"
                                                aria-live="polite"
                                                class="default text"
                                                role="alert"
                                              >
                                                Add tags...
                                              </div>
                                              <i
                                                aria-hidden="true"
                                                class="dropdown icon"
                                              />
                                              <div
                                                aria-multiselectable="true"
                                                class="menu transition"
                                                role="listbox"
                                              >
                                                <div
                                                  class="message"
                                                >
                                                  No results found.
                                                </div>
                                              </div>
                                            </div>,
                                          }
                                        }
                                      >
                                        <div
                                          aria-expanded={false}
                                          className="ui fluid multiple search selection dropdown"
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          onClick={[Function]}
                                          onFocus={[Function]}
                                          onMouseDown={[Function]}
                                          role="combobox"
                                        >
                                          <Ref
                                            innerRef={
                                              Object {
                                                "current": <input
                                                  aria-autocomplete="list"
                                                  autocomplete="off"
                                                  class="search"
                                                  tabindex="0"
                                                  type="text"
                                                  value=""
                                                />,
                                              }
                                            }
                                          >
                                            <RefFindNode
                                              innerRef={
                                                Object {
                                                  "current": <input
                                                    aria-autocomplete="list"
                                                    autocomplete="off"
                                                    class="search"
                                                    tabindex="0"
                                                    type="text"
                                                    value=""
                                                  />,
                                                }
                                              }
                                            >
                                              <DropdownSearchInput
                                                autoComplete="off"
                                                key="text"
                                                onChange={[Function]}
                                                style={
                                                  Object {
                                                    "width": undefined,
                                                  }
                                                }
                                                tabIndex={0}
                                                type="text"
                                                value=""
                                              >
                                                <input
                                                  aria-autocomplete="list"
                                                  autoComplete="off"
                                                  className="search"
                                                  onChange={[Function]}
                                                  style={
                                                    Object {
                                                      "width": undefined,
                                                    }
                                                  }
                                                  tabIndex={0}
                                                  type="text"
                                                  value=""
                                                />
                                              </DropdownSearchInput>
                                            </RefFindNode>
                                          </Ref>
                                          <span
                                            className="sizer"
                                          />
                                          <div
                                            aria-atomic={true}
                                            aria-live="polite"
                                            className="default text"
                                            role="alert"
                                          >
                                            Add tags...
                                          </div>
                                          <Icon
                                            as="i"
                                            className=""
                                            name="dropdown"
                                            onClick={[Function]}
                                          >
                                            <i
                                              aria-hidden="true"
                                              className="dropdown icon"
                                              onClick={[Function]}
                                            />
                                          </Icon>
                                          <DropdownMenu
                                            aria-multiselectable={true}
                                            role="listbox"
                                          >
                                            <div
                                              aria-multiselectable={true}
                                              className="menu transition"
                                              role="listbox"
                                            >
                                              <div
                                                className="message"
                                              >
                                                No results found.
                                              </div>
                                            </div>
                                          </DropdownMenu>
                                        </div>
                                      </RefFindNode>
                                    </Ref>
                                  </Dropdown>
                                </div>
                              </GridRow>
                              <GridRow>
                                <div
                                  className="row"
                                >
                                  <Button
                                    as="button"
                                    disabled={true}
                                    onClick={[Function]}
                                    primary={true}
                                  >
                                    <Ref
                                      innerRef={
                                        Object {
                                          "current": <button
                                            class="ui primary disabled button"
                                            disabled=""
                                            tabindex="-1"
                                          >
                                            <span>
                                              Save
                                            </span>
                                          </button>,
                                        }
                                      }
                                    >
                                      <RefFindNode
                                        innerRef={
                                          Object {
                                            "current": <button
                                              class="ui primary disabled button"
                                              disabled=""
                                              tabindex="-1"
                                            >
                                              <span>
                                                Save
                                              </span>
                                            </button>,
                                          }
                                        }
                                      >
                                        <button
                                          className="ui primary disabled button"
                                          disabled={true}
                                          onClick={[Function]}
                                          tabIndex={-1}
                                        >
                                          <FormattedMessage
                                            defaultMessage="Save"
                                            id="app.program_tags.save"
                                            values={Object {}}
                                          >
                                            <span>
                                              Save
                                            </span>
                                          </FormattedMessage>
                                        </button>
                                      </RefFindNode>
                                    </Ref>
                                  </Button>
                                </div>
                              </GridRow>
                            </div>
                          </Grid>
                        </ProgramTags>
                      </Connect(ProgramTags)>
                    </InjectIntl(Connect(ProgramTags))>
                  </withCookies(Component)>
                </div>
              </GridRow>
              <Divider>
                <div
                  className="ui divider"
                />
              </Divider>
              <GridRow>
                <div
                  className="row"
                >
                  <Segment
                    basic={true}
                    compact={true}
                  >
                    <div
                      className="ui basic compact segment"
                    >
                      <Component>
                        <div />
                      </Component>
                    </div>
                  </Segment>
                </div>
              </GridRow>
            </div>
          </GridColumn>
          <GridColumn
            width={10}
          >
            <div
              className="ten wide column"
            >
              <Component
                location={
                  Object {
                    "state": Object {
                      "readOnly": false,
                    },
                  }
                }
              >
                <div />
              </Component>
            </div>
          </GridColumn>
          <GridColumn
            width={3}
          >
            <div
              className="three wide column"
            >
              <GridRow>
                <div
                  className="row"
                >
                  <Header
                    as="h2"
                    textAlign="center"
                  >
                    <h2
                      className="ui center aligned header"
                    >
                      <FormattedMessage
                        defaultMessage="Rovers"
                        id="app.mission_control.rovers"
                        values={Object {}}
                      >
                        <span>
                          Rovers
                        </span>
                      </FormattedMessage>
                    </h2>
                  </Header>
                  <Segment
                    raised={true}
                    style={
                      Object {
                        "margin": "10px",
                        "maxHeight": 200,
                        "overflow": "auto",
                      }
                    }
                  >
                    <div
                      className="ui raised segment"
                      style={
                        Object {
                          "margin": "10px",
                          "maxHeight": 200,
                          "overflow": "auto",
                        }
                      }
                    >
                      <Connect(RoverConnection)>
                        <RoverConnection
                          changeLeftSensorState={[Function]}
                          changeRightSensorState={[Function]}
                          commands={Array []}
                          connectToRover={[Function]}
                          disconnectFromRover={[Function]}
                          isFetching={false}
                          rover={null}
                          rovers={
                            Object {
                              "next": null,
                              "previous": null,
                              "results": Array [],
                            }
                          }
                          scanForRover={[Function]}
                        >
                          <Button
                            as="button"
                            onClick={[Function]}
                            primary={true}
                          >
                            <Ref
                              innerRef={
                                Object {
                                  "current": <button
                                    class="ui primary button"
                                  >
                                    <span>
                                      Connect to rover
                                    </span>
                                  </button>,
                                }
                              }
                            >
                              <RefFindNode
                                innerRef={
                                  Object {
                                    "current": <button
                                      class="ui primary button"
                                    >
                                      <span>
                                        Connect to rover
                                      </span>
                                    </button>,
                                  }
                                }
                              >
                                <button
                                  className="ui primary button"
                                  onClick={[Function]}
                                >
                                  <FormattedMessage
                                    defaultMessage="Connect to rover"
                                    id="app.rover_list.connect"
                                    values={Object {}}
                                  >
                                    <span>
                                      Connect to rover
                                    </span>
                                  </FormattedMessage>
                                </button>
                              </RefFindNode>
                            </Ref>
                          </Button>
                        </RoverConnection>
                      </Connect(RoverConnection)>
                    </div>
                  </Segment>
                </div>
              </GridRow>
              <Divider>
                <div
                  className="ui divider"
                />
              </Divider>
              <GridRow>
                <div
                  className="row"
                >
                  <Header
                    as="h2"
                    textAlign="center"
                  >
                    <h2
                      className="ui center aligned header"
                    >
                      <FormattedMessage
                        defaultMessage="Sensors"
                        id="app.mission_control.sensors"
                        values={Object {}}
                      >
                        <span>
                          Sensors
                        </span>
                      </FormattedMessage>
                    </h2>
                  </Header>
                  <Segment
                    raised={true}
                    style={
                      Object {
                        "margin": "10px",
                      }
                    }
                  >
                    <div
                      className="ui raised segment"
                      style={
                        Object {
                          "margin": "10px",
                        }
                      }
                    >
                      <Component>
                        <div />
                      </Component>
                    </div>
                  </Segment>
                </div>
              </GridRow>
              <Divider>
                <div
                  className="ui divider"
                />
              </Divider>
              <GridRow>
                <div
                  className="row"
                >
                  <Header
                    as="h2"
                    textAlign="center"
                  >
                    <h2
                      className="ui center aligned header"
                    >
                      <FormattedMessage
                        defaultMessage="Debug Console"
                        id="app.mission_control.console"
                        values={Object {}}
                      >
                        <span>
                          Debug Console
                        </span>
                      </FormattedMessage>
                    </h2>
                  </Header>
                  <Segment
                    style={
                      Object {
                        "margin": "10px",
                      }
                    }
                  >
                    <div
                      className="ui segment"
                      style={
                        Object {
                          "margin": "10px",
                        }
                      }
                    >
                      <Component>
                        <div />
                      </Component>
                    </div>
                  </Segment>
                </div>
              </GridRow>
            </div>
          </GridColumn>
        </div>
      </GridRow>
    </div>
  </Grid>
</MissionControl>
`;
